OSDOCS#10223: Consolidate "Installing from OperatorHub" procs & Single vs AllNamespace improvements (#79401)
openshift/openshift-docs: modules/olm-installing-from-operatorhub-using-cli.adoc

:_mod-docs-content-type: PROCEDURE
[id="olm-installing-operator-from-operatorhub-using-cli_{context}"]
= Installing from OperatorHub by using the CLI

Instead of using the {product-title} web console, you can install an Operator from OperatorHub by using the CLI. Use the `oc` command to create or update a `Subscription` object.

For `SingleNamespace` install mode, you must also ensure an appropriate Operator group exists in the related namespace. An Operator group, defined by an `OperatorGroup` object, selects target namespaces in which to generate required RBAC access for all Operators in the same namespace as the Operator group.

[TIP]
====

> For most usage, the web console method of this procedure is preferred because it automates tasks in the background, such as handling the creation of `OperatorGroup` and `Subscription` objects automatically when choosing `SingleNamespace` mode.

Reviewer: I'm not sure _usage_ is the right word. I'm prolly wrong, so feel free to ignore.

```suggestion
In most cases, the web console method of this procedure is preferred because it automates tasks in the background, such as handling the creation of `OperatorGroup` and `Subscription` objects automatically when choosing `SingleNamespace` mode.
```

$ oc describe packagemanifests <operator_name> -n openshift-marketplace

> ----

Reviewer: Output showing what to look for?

```
----
Name:         etcd
Namespace:    openshift-marketplace
Labels:       catalog=community-operators
              catalog-namespace=openshift-marketplace
              operatorframework.io/arch.amd64=supported
              operatorframework.io/os.linux=supported
              provider=CNCF
              provider-url=
Annotations:  <none>
API Version:  packages.operators.coreos.com/v1
Kind:         PackageManifest
# ...
      Display Name:  etcd
      Install Modes:
        Supported:  true
        Type:       OwnNamespace
        Supported:  true <1>
        Type:       SingleNamespace
        Supported:  false
        Type:       MultiNamespace
        Supported:  false
        Type:       AllNamespaces
----
<1> Specifies that the Operator supports `SingleNamespace` install mode.
```

. If the Operator you intend to install uses the `AllNamespaces` install mode, skip this step, because the `openshift-operators` namespace already has an appropriate Operator group in place by default, called `global-operators`.
+

> If the Operator you intend to install uses the `SingleNamespace` install  mode and you do not already have an appropriate Operator group in place, you must create one by following these steps:

Reviewer: What if the Operator supports both `AllNamespaces` and `SingleNamespace` install modes?

```
      Display Name:  Couchbase Operator
      Install Modes:
        Supported:  true
        Type:       OwnNamespace
        Supported:  true
        Type:       SingleNamespace
        Supported:  false
        Type:       MultiNamespace
        Supported:  true
        Type:       AllNamespaces
```

Author: It's more about which of the supported methods you are choosing to utilize during this particular installation; I will sharpen this up, thx!

+
--
.. An Operator group, defined by an `OperatorGroup` object, selects target namespaces in which to generate required RBAC access for all Operators in the same namespace as the Operator group.
+
Create an `OperatorGroup` object YAML file, for example `operatorgroup.yaml`, for `SingleNamespace` install mode:
+
.Example `OperatorGroup` object for `SingleNamespace` install mode
[source,yaml]
----
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: <operatorgroup_name>
  namespace: <namespace> <1>
spec:
  targetNamespaces:
  - <namespace> <1>
----
<1> For `SingleNamespace` install mode, use the same `<namespace>` value for both the `metadata.namespace` and `spec.targetNamespaces` fields.

.. Create the `OperatorGroup` object:
+
[source,terminal]
----
$ oc apply -f operatorgroup.yaml
----
--
+
[NOTE]
====

> * The web console version of this procedure handles the creation of the `OperatorGroup` and `Subscription` objects automatically in the background for you when choosing `SingleNamespace` mode.

Reviewer: Not to be flippant, but why do I care what the web console version of this procedure does? Seems as though this would be better in the web console procedures.

Author: I think it was initially added because we primarily recommend users go with the web console method since this Operator group stuff can get tricky when manually creating.

Author: Moved this to the intro and gave it more context wrt preferred method.

+
At this point, OLM is now aware of the selected Operator. A cluster service version (CSV) for the Operator should appear in the target namespace, and APIs provided by the Operator should be available for creation.

.Verification

. Check the status of the `Subscription` object for your installed Operator:
+
[source,yaml]
----

> $ oc describe subscription <subscription_name>

Reviewer 2: It's better to specify the namespace here.

----

. If you created an Operator group for `SingleNamespace` install mode, check the status of the `OperatorGroup` object:
+
[source,yaml]
----

> $ oc describe operatorgroup <operatorgroup_name>

Reviewer 2: Same, it's better to specify the namespace here.

<3> Include the subscription ID.
--

> .. Create the `Subscription` object:

Reviewer: ```suggestion
.. Create the `Subscription` object by running the following command:
```

+
[source,terminal]
----
$ oc apply -f subscription.yaml
----

> . If you set the `installPlanApproval` field to `Manual`, manually approve the pending install plan to complete the Operator installation.

Reviewer: >  manually approve the pending install plan

I assume it will be obvious how to manually approve?

Author: It was linked in "Additional resources", but I've added the following sentence to make that connection more clear:
> For more information, see "Manually approving a pending Operator update".

At this point, OLM is now aware of the selected Operator. A cluster service version (CSV) for the Operator should appear in the target namespace, and APIs provided by the Operator should be available for creation.

.Verification

> . Check the status of the `Subscription` object for your installed Operator:

Reviewer: ```suggestion
. Check the status of the `Subscription` object for your installed Operator by running the following command:
```

+
[source,terminal]
----
$ oc describe subscription <subscription_name> -n <namespace>
----

> . If you created an Operator group for `SingleNamespace` install mode, check the status of the `OperatorGroup` object:

Reviewer: ```suggestion
. If you created an Operator group for `SingleNamespace` install mode, check the status of the `OperatorGroup` object by running the following command:
```
